tempodb/compactor.go

```diff
@@ -380,3 +422,21 @@ func (i instrumentedObjectCombiner) Combine(dataEncoding string, objs ...[]byte)
 	}
 	return b, wasCombined, err
 }
+
+// doForAtLeast executes the function f. It blocks for at least the passed duration but can go longer. if context is cancelled after
+// the function is done we will bail immediately. in the current use case this means that the process is shutting down
+// we don't force f() to cancel, we assume it also responds to the cancelled context
+func doForAtLeast(ctx context.Context, dur time.Duration, f func()) {
```

Comment:
Hmm, I'm trying to understand this.  It looks like this still only executes `f` once no matter the context status.  What am I overlooking?

Reply:
Oh we're holding the blocked call to f for at least duration.

Reply:
it was meant to replicate the current logic in a more straight forward way. i'm not 100% sure, but the current logic:
```
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}

		select {
		case <-ticker.C:
			rw.doCompaction(ctx)
		case <-ctx.Done():
			return
		}
	}
```

Basically just attempts to run the `doCompaction` function once every `CompactionCycle`. Also, if compaction takes a long time then the ticker would have elapsed potentially hundreds of times over. I wonder if this is related to our go 1.23 timer panic?

So I tried to write a simpler method that:
- ran the compaction function
- if it took more than the CompactionCycle then immediately return so we can do another
- if it took less time than the CompactionCycle then wait for the difference while also watching for a context cancel.

Reply:
This is clever and I like the approach.  It took a minute to see the reason, but holding a wait period for the cancellation makes sense.